### PR TITLE
shell: account expiration and password expiration

### PR DESCRIPTION
**Pull request by fridex**
Based on https://github.com/cockpit-project/cockpit/issues/644

### CODE REVIEW

**Comment by andreasn** [comment]
I was going to try this one out again today, but I think it bitrotted :/
Would you be able to rebase it on top of master?

**Comment by fridex** [comment]
Rebased. It's been some time, hopefully I didn't break something.

**Comment by andreasn** [comment]
I created a user and selected "Force user to change password on next login", but when I tried to log in with that user I got "Authentication token is no longer valid; new one required"

The "Force user to change password on next login" is also a very long label. I would expect it to become even longer in some translations, so this needs to be shortened.

**Comment by andreasn** [comment]
I think it will need to throw up some kind of dialog to change the password (and only that?). Allowing you to set it there, and then bring you back to the login screen.

**Comment by andreasn** [comment]
Some previous art from FreeIPA

screenshot from 2015-06-24 15-57-07
screenshot from 2015-06-24 15-56-22

**Comment by andreasn** [comment]
When the user account expires, does that mean that it gets deleted, or that the account gets locked out?

**Comment by fridex** [comment]
> When the user account expires, does that mean that it gets deleted, or that the account gets locked out?

Locking an account means disabling user's password (the user is unable to log in). When an account is expired, user is notified after entering a password (and logged out). "usermod" is used, you can find details in "man usermod"

**Comment by andreasn** [comment]
Instead of "Force user to...." I would suggest the string "Require password change on next login". Force feels too aggressive.

**Comment by andreasn** [comment]
Is the number 99999 days the system default for "Number of days between password change", or is that something custom that's part of your pull request?

**Comment by stefwalter** [comment]
> I think it will need to throw up some kind of dialog to change the password (and only that?). Allowing you to set it there, and then bring you back to the login screen.

The login screen would need to adapt, change its text to ask for a new password to be entered twice. You can see that Windows does this. I think GDM might do this too.

This happens at the PAM level. Therefore we can't log in or show any of the rest of Cockpit to do this. The logic to make this happen would need to happen at the login screen level, and related auth code in cockpit-ws. This part of the feature should also work with both cockpit-session and sshd logins in order to make it mergeable.

**Comment by andreasn** [comment]
also, is that 99999 days starting with today as 0?

**Comment by andreasn** [comment]
> The login screen would need to adapt, change its text to ask for a new password to be entered twice. You can see that Windows does this. I think GDM might do this too.

I've checked out what GDM did. Will create some mockups for this.

**Comment by andreasn** [comment]
Mockup for setting the password at the login screen:
Mockup

**Comment by stefwalter** [comment]
So this pull request looks pretty dead. The blocker is that we don't support account expiration (which makes less sense for admins anyway) on our login screen, so this is incomplete as a feature.

But if we reduce scope, and just **show** the account and expiration information, then it could be considered complete.

**Comment by andreasn** [comment]
> But if we reduce scope, and just show the account and expiration information, then it could be considered complete.

Sure, I think that could work.

**Comment by stefwalter** [comment]
https://trello.com/c/7d5tgGMJ/264-atomic-visual-boot-force-password-reset-cockpit

**Comment by andreasn** [comment]
"Your password HAS expired" since it's an "it".

**Comment by stefwalter** [comment]
Now that prompting for expired passwords is in Cockpit, this is unblocked and we can finish the users page parts of this.

**Comment by stefwalter** [comment]
@fridex Could you enable me to work on this by checking the "Allow edits from maintainers" checkbox?

**Comment by fridex** [comment]
@stefwalter sure! checked.

**Comment by stefwalter** [comment]
@garrett or @andreasn Here's screenshots if you want to give advice on how this is laid out, wording, or design:

screenshot from 2017-07-24 10-31-29
screenshot from 2017-07-24 10-31-37
screenshot from 2017-07-24 10-31-40
screenshot from 2017-07-24 10-31-48
screenshot from 2017-07-24 10-31-51
screenshot from 2017-07-24 10-31-54
screenshot from 2017-07-24 10-31-57

**Comment by andreasn** [comment]
From the screenshots, this looks good to me in general. I would only fix a couple of controls and some wording. I'll get some wireframes shortly.

When a password expires, you'll need to type a new password, right? What happens when an account expires?

**Comment by stefwalter** [comment]
> What happens when an account expires?

The account can no longer log in.

File: pkg/users/local.js [hunk 1]
```diff
@@ -1057,6 +1126,136 @@ function PageAccountConfirmDelete() {
...
+        } else {
+            date = null;
+        }
+
+        /* Actually perform the action if valid */
+        var prog = ["/usr/sbin/usermod", "-e"];
+        var account_id = $("#account-expiration").data("account-id");
+        if (!promise) {
+            if (date)
+                prog.push(date.toISOString().substr(0, 10));
```
Comment (larskarlitski) [inline pkg/users/local.js]: This, in combination with `new Date` further up, depends on the Browser's time zone. I'm in CEST and the expiration date is always set to one day earlier than the day I entered in the dialog.
```js
(new Date('Jul 20, 2017')).toISOString().substr(0, 10)
// -> "2017-07-19"
```

File: pkg/users/local.js [hunk 2]
```diff
@@ -1057,6 +1126,136 @@ function PageAccountConfirmDelete() {
...
+            else
+                prog.push("");
+            prog.push(account_id);
+            console.log(prog);
```
Comment (stefwalter) [inline pkg/users/local.js]: Removed.

File: pkg/users/local.js (hunk 1 above)
Comment (stefwalter) [inline pkg/users/local.js]: Fixed this. We display dates on the main page exactly as chage rendered them. Editable dates are always displayed as ISO 8601 ... and the datepicker is updated to make that the case. Fixed parsing and avoiding interpreting local time zones.

**Comment by andreasn** [comment]
mockup

I wanted to address a couple of things in this design:
 * Explain what will happen when an account or password expire (I was a bit unsure myself if the account expiration would delete or lock the account)
 * Make it radio buttons rather than checkbox, to show the two different states
 * Reuse the wording "Lock account" and "require password change" as much as possible, since we have those concepts already. Still didn't manage to fully get rid of the word "expire", but hopefully this is good enough.

Thanks for the feedback from @garrett on IRC!

**Comment by andreasn** [comment]
Note that the string "Never lock account" would carry out to the main page as well.

**Comment by stefwalter** [comment]
Updated for @andreasn and @larskarlitski design and review.

**Comment by andreasn** [comment]
Works a lot better now!
I realized it is possible to lock the account in the past. Won't cause harm I think, but a bit odd. Can be a followup fix.

**Comment by stefwalter** [comment]
> I realized it is possible to lock the account in the past. 

That's something Linux (chage command) allows in general. I think we don't need to worry about that.

**Comment by larskarlitski** [comment]
Approved, thanks for the changes.

I'm still wondering about the weird error message on the login screen, but that's not coming from this patch.

**Pull request merged**